packages/ai-core/src/browser/frontend-prompt-customization-service.ts

```diff
@@ -72,55 +104,259 @@ export class FrontendPromptCustomizationServiceImpl implements PromptCustomizati
 
     protected async update(): Promise<void> {
         this.toDispose.dispose();
-        // we need to assign a local variable, so that updates running in parallel don't interfere with each other
-        const _templates = new Map<string, string>();
-        this.templates = _templates;
-        this.trackedTemplateURIs.clear();
+        // we need to assign local variables, so that updates running in parallel don't interfere with each other
+        const templatesMap = new Map<string, TemplateEntry>();
+        const trackedURIs = new Set<string>();
+        const loadedTemplates = new Map<string, TemplateEntry>();
+        const watchedFilesMap = new Map<string, WatchedFileInfo>();
 
+        // Process in order of priority (lowest to highest)
+        // First process the main template directory (lowest priority)
         const templateURI = await this.getTemplatesDirectoryURI();
+        await this.processTemplateDirectory(templatesMap, trackedURIs, loadedTemplates, templateURI, TemplatePriority.GLOBAL_TEMPLATE_DIR);
+
+        // Process additional template directories (medium priority)
+        for (const dirPath of this.additionalTemplateDirs) {
+            const dirURI = URI.fromFilePath(dirPath);
+            await this.processTemplateDirectory(templatesMap, trackedURIs, loadedTemplates, dirURI, TemplatePriority.ADDITIONAL_TEMPLATE_DIR);
+        }
+
+        // Process specific template files (highest priority)
+        await this.processTemplateFiles(templatesMap, trackedURIs, loadedTemplates, watchedFilesMap);
+
+        this.effectiveTemplates = templatesMap;
+        this.trackedTemplateURIs = trackedURIs;
+        this.allLoadedTemplates = loadedTemplates;
+        this.watchedFiles = watchedFilesMap;
+
+        this.onDidChangeCustomAgentsEmitter.fire();
+    }
+
+    /**
+     * Adds a template to the templates map, handling conflicts based on priority
+     * @param templatesMap The map to add the template to
+     * @param id The template ID
+     * @param content The template content
+     * @param priority The template priority
+     * @param sourceUri The URI of the source file (used to distinguish updates from conflicts)
+     * @param loadedTemplates The map to track all loaded templates
+     */
+    protected addTemplate(
+        templatesMap: Map<string, TemplateEntry>,
+        id: string,
+        content: string,
+        priority: number,
+        sourceUri: string,
+        loadedTemplates: Map<string, TemplateEntry> = this.allLoadedTemplates
```

Comment:
This should not be an optional parameter. If we refer to `this.allLoadedTemplates` while loading, we might run into race conditions (interleaving updates) again.